### ISSUE TITLE
Add possibility to consistently get object property name

### ISSUE THREAD

**Issue by DisDis** [dart-lang/language#251]
We need to get object property name in cases like Angular 2 ngOnChanges (when you need to compare key from changes map with object property name):

`ngOnChanges(Map<String, SimpleChange> changes);`
It would be great to have something like [nameof](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/nameof) in C#, so that we could get property name and don't be afraid that someone will rename property.

`nameof(obj.propertyName); // 'propertyName'`

**Comment by eernstg** [comment]
I guess the desired guarantee here (cf. the C# [nameof](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/nameof)) would be that a given symbol corresponding to a single identifier or operator does in fact exist as the name of a declared member of some type.

If we make this a built-in property of certain symbols (say, by spelling them as `##foo` rather than `#foo`), it is not immediately obvious to me how we could support specification of that type. We could of course just require that there is some type, somewhere, that has a member called `foo`, and if there is no such type then `##foo` is a compile-time error.

But that's a rather weak guarantee. So maybe the following would be more practical:

```dart
main() {
  var s1 = memberSymbol<A>(#foo); // OK.
  var s2 = memberSymbol<B>(#bar); // OK.
  var s2 = memberSymbol<B>(#foo); // OK, it's in the interface.
  var s3 = memberSymbol<A>(#baz, staticMember: true); // OK.
}

abstract class A {
  foo();
  Symbol get test => memberSymbol(#foo); // OK, type argument `A` inferred.
}

class B extends /*or implements, or mixes in*/ A {
  int get bar => 42;
  static void baz() {}
}
```

This would implementations of `noSuchMethod` to rely on inference (the desired and inferred type argument is always the enclosing class), and all other usages could give an explicit type argument if needed.

The `memberSymbol` "function" would be special in that an application of it is a constant expression, and the evaluation just returns a correct value (that is, something which is a constant symbol such that the given type does have a member with that name), and an incorrect value (not a member) causes a compile-time error.

We could extend this idea slightly to allow for an optional argument `staticMember` which would allow us to check that a given symbol is the name of a static member of the given type (which must then be a class type, of course); we could also just omit support for that kind of check, or we could allow it without any special marker.

**Comment by Renesanse** [comment]
Any news guys?

**Comment by leafpetersen** [comment]
> Any news guys?

I'd say that this request is still open, and seems to have some level of support, but there are a lot of higher priority things occupying both the language team and the implementation teams right now, so I don't expect this to bubble to the top of the queue in the short term.

**Comment by tomdowdell** [comment]
The C# ```nameof()``` operator is definitely useful for those of us who create professional developer tools. Please add this feature!

The following example is not trying to prove the benefit, it just shows a use code for a minimal implementation that would satisfy my primary needs:

```
class Ant {
  String firstName;

  Bat() {
  var s = nameof(firstName);
  }
}
```

**Comment by raulchall** [comment]
+1 to this, just couple days into Flutter + Firestore started looking for this

**Comment by rrousselGit** [comment]
Maybe an alternate solution is to create a new lint rule that warns when a symbol's name doesn't match any variable available?

**Comment by festelo** [comment]
There can be a problem with code minification. Imagine we have client-server app, client and server are both on dart. We also have some shared code that used by both.

Shared code:
'''
class SomeModel {
    final name;
    
    SomeModel.fromMap(map):
        name = map[nameof(name)];

    toMap() => {
        nameof(name): name
    };
}
'''

Flutter web application **built in release** will not be able to deserialize code serialized by Server and vice versa.

Client's toMap will produce map 
```
{ 'mF': 'some name' }
```

Server's toMap will produce map
'''
{'name': 'some name'}
'''

Because type and field names are minified in flutter web when built in release, but they are not minified in server.

**Comment by zs-dima** [comment]
`nameof(obj.propertyName); // 'propertyName'` - this feature more then welcome

**Comment by esDotDev** [comment]
Fwiw, this would be nice for the new Restorable API in flutter, so we don't need to constantly repeat ourselves:
```
RestorableInt someInt = RestorableInt(0);
...
@override
  void restoreState(RestorationBucket oldBucket, bool initialRestore) {
    registerForRestoration(someInt, "someInt"); // "someInt" is redundant here most of the time 
  }
```
The id in this case is a string that is supposed to be unique within the scope of a class, which is of course accomplished by just using it's property name;

**Comment by tatumizer** [comment]
> registerForRestoration(someInt, "someInt");
   //registerForRestoration(someInt); // This should be enough, and is less prone to bugs

This won't be enough. Consider an example:
```
void foo(int x) {
  print("${nameof(x)}=$x"); 
}
var amount=1000;
foo(amount);
```
This program will print "x=1000", not "amount=1000" as you expect.
To pass the real name of a variable, you still need 2 parameters:
```
void foo(int x, String varName) {
  print("$varName=$x"); 
}
var amount=1000;
foo(amount, nameof(amount));
```

**Comment by esDotDev** [comment]
Right, thanks for the correction. The longer method is still preferred as it's avoids the magic strings and lets the compiler keep the names in sync.

**Comment by esDotDev** [comment]
Another real world use case we came across, We're trying to come up with some concise method of re-using state in a StatelessWidget in Flutter.

Landed on an API that needs a unique-in-scope string, very similar to Restoration:
```
StatefulAnimation anim1 = getProp("anim1", create: (){ ... })
```
I think `getProp(nameOf(anim1), ...)` would be preferred here.

**Comment by bannmann** [comment]
Another use case which was not yet mentioned is throwing exceptions that include the parameter name:

```
set foo(String value) {
  _foo = ArgumentError.checkNotNull(value, nameof(value));
}

void bar(String data) {
  if (!isValid(data)) {
    throw ArgumentError.value(data, nameof(data));
  }
  ...
}

void quux(String data) {
  if (data == null) {
    throw ArgumentError.notNull(nameof(data));
  }
  ...
}
```

**Comment by shtse8** [comment]
> > There can be a problem with code minification. Imagine we have client-server app, client and server are both on dart. We also have some shared code that used by both.
> 
> Wouldn't nameof() be resolved before any minification, so property names would still match the keys in the serialized data?

before minification would make it to be a const string, minification wouldn't take affect on that.

**Comment by codelovercc** [comment]
A `nameof` expression produces the name of a variable, type, or member as the string constant. A `nameof` expression is evaluated at compile time and has no effect at run time.  https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/nameof
This is not about reflection and don't take affect on minification.
validation error response on LoginModel: {"title":"validation error","errors":{"account":["Account is required."],"password":["Passowrd is required"]}} 
```dart
class LoginModelProblemDetails {
  final List<String>? account;

  final List<String>? password;

  final List<String>? accountType;

  LoginModelProblemDetails({
    this.account,
    this.password,
    this.accountType,
  });

  bool hasAnyError(){
    return account?.isNotEmpty == true || password?.isNotEmpty == true || accountType?.isNotEmpty == true;
  }
  
  bool hasError(String fieldName){
    switch(fieldName){
      case "account":
        return account?.isNotEmpty == true;
      case "password":
        return password?.isNotEmpty == true;
      case "accountType":
        return accountType?.isNotEmpty == true;
      default:
        return false;
    }
  }
```
If I can use `nameof` keyword, that would be

```dart
class LoginModelProblemDetails {
  final List<String>? account;

  final List<String>? password;

  final List<String>? accountType;

  LoginModelProblemDetails({
    this.account,
    this.password,
    this.accountType,
  });

  bool hasAnyError(){
    return account?.isNotEmpty == true || password?.isNotEmpty == true || accountType?.isNotEmpty == true;
  }
  
  bool hasError(String fieldName){
    switch(fieldName){
      case nameof(account):
        return account?.isNotEmpty == true;
      case nameof(password):
        return password?.isNotEmpty == true;
      case nameof(accountType):
        return accountType?.isNotEmpty == true;
      default:
        return false;
    }
  }
```
check it:
```dart
final problemDetails = LoginModelProblemDetails.fronJson(jsonDecode('{"title":"validation error","errors":{"account":["Account is required."],"password":["Passowrd is required"]}} '));
print(problemDetails.hasError(nameof(account));
```
I have a lot of models `ModelProblemDetails` like, and it aways decode from json that the server responded, my UI layer use method `hasError` to toggle the validation state that display to user. I need keyword `nameof` to save my life.

**Comment by bdemeure** [comment]
+1 for nameof() equivalent that is 'resolved' before minification

currently we create manualy some static const that matches property names... but error can easily occurs if const value is not spelled exactly as the property

**Comment by aa2013** [comment]
+1
